### PR TITLE
Issue 2111 custom transport efficiency

### LINKED ISSUE #2111: Adding customized function for tortuosity
### Description

I was wondering if it is possible to enable adding customized non-Bruggeman functions to define tortuosity or MacMullin number as a function of porosity?

Thanks. 

### Motivation

_No response_

### Possible Implementation

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by TomTranter**
# Description

Adding new methods for calculating the transport efficiency beyond Bruggeman coefficient. This includes using the following relations for tortuosity factor ($\tau$) for use in transport efficiency $B=\epsilon/\tau$:

image

- Tortuosity factor: allows direct specification of a tortuosity factor
These were taken from 10.1016/j.ces.2007.03.041 and http://dx.doi.org/10.1016/j.coche.2016.02.006

Fixes #2111 

#

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all`
- [x] The documentation builds: `$ python run-tests.py --doctest`

You can run unit and doctests together at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by TomTranter** [comment]
This line looks to be wrong as active material won't always be the only solid conducting material. Using 1-porosity would be better
https://github.com/pybamm-team/PyBaMM/blob/bd5e240a26927800fd7338da294915be6ca7966f/pybamm/models/submodels/transport_efficiency/bruggeman_transport_efficiency.py#L39

File: pybamm/models/submodels/transport_efficiency/general_transport_efficiency.py
Comment (rtimms) [inline pybamm/models/submodels/transport_efficiency/general_transport_efficiency.py:24]: can Bruggeman and tortuosity factor go in this method too, or are they treated differently for some reason?

Comment (TomTranter) [reply pybamm/models/submodels/transport_efficiency/general_transport_efficiency.py:24]: I don't think it reduces total lines of code as you would have to do the if statement on the domain inside the sub-function

Comment (TomTranter) [reply pybamm/models/submodels/transport_efficiency/general_transport_efficiency.py:24]: They are using additional specific parameters whereas the other ones just use porosity

Comment (rtimms) [reply pybamm/models/submodels/transport_efficiency/general_transport_efficiency.py:24]: Makes sense, got it

**Comment by TomTranter** [comment]
@rtimms @brosaplanella this should be ready to go. Benchmark regressions appear to be random and unrelated to changes.

File: pybamm/models/submodels/transport_efficiency/general_transport_efficiency.py
Comment (brosaplanella) [inline pybamm/models/submodels/transport_efficiency/general_transport_efficiency.py:8]: Do we anticipate transport efficiency models that would not be into `general_transport_efficiency.py`? In that case, it may be better to get rid of the base class and just have a `transport_efficiency.py` file. Alternatively, we could split `general_transport_efficiency.py` into subclasses (one for each model). Not sure what's best practice in these situations honestly, so happy to be convinced either way.

Comment (brosaplanella) [reply pybamm/models/submodels/transport_efficiency/general_transport_efficiency.py:8]: One advantage of the latter is that when checking the submodel components it will be a lot more explicit.

**Review by brosaplanella — APPROVED** [review]
Looks good, just see my comment on the classes. We should wait to merge until we have done the rc.

File: pybamm/models/submodels/transport_efficiency/general_transport_efficiency.py
Comment (rtimms) [reply pybamm/models/submodels/transport_efficiency/general_transport_efficiency.py:8]: For kinetics, each different option is its own class that just implements `_get_kinetics`

File: pybamm/models/submodels/transport_efficiency/bruggeman.py
Comment (brosaplanella) [inline pybamm/models/submodels/transport_efficiency/bruggeman.py:9]: Thanks, I like it much better now! Could you add a docstring with the specific model name for each class and, if possible, the reference?

**Review by brosaplanella — APPROVED** [review]
Looks great, thanks! Just a small comment on the docstrings. Let's hold this until the rc is published.

File: CHANGELOG.md
Comment (Saransh-cpp) [inline CHANGELOG.md:6]: This should be moved to the unreleased section.

Comment (TomTranter) [reply CHANGELOG.md:8]: Done

**Comment by TomTranter** [comment]
@brosaplanella macOS builds failing but probably nothing to do with me

**Review by agriyakhetarpal — COMMENTED** [review]
Hello @TomTranter! Could you add references to the docstrings of the model classes using the BibTeX directly instead, e.g.,

```python
:footcite:t:`bruggeman1935berechnung`
```
and so on?

It's much easier than hard-coding the title of the paper, plus the `sphinxcontrib-bibtex` extension will automatically parse it and add the references to the documentation.

File: pybamm/models/submodels/transport_efficiency/tortuosity_factor.py
Comment (brosaplanella) [inline pybamm/models/submodels/transport_efficiency/tortuosity_factor.py:9]:
```suggestion
    """Submodel for user supplied tortuosity factor transport_efficiency
```

**Review by brosaplanella — APPROVED** [review]
Looks good to me, just a small typo on a docstring. Happy to merge when that's suggestion is merged.

**Comment by brosaplanella** [comment]
Who was in the Hackathon team, so we can add them in the contributors table?

**Comment by TomTranter** [comment]
> Who was in the Hackathon team, so we can add them in the contributors table?

@juliawind @amirDahari1 @tommaull @Ruimin-S @isaacsquires are the ones with code commits

**Comment by brosaplanella** [comment]
The doc tests are failing and I think this blocks merging the PR. @agriyakhetarpal can you help with this? It seems the issue is with the references.

File: docs/source/api/models/submodels/transport_efficiency/index.rst
Comment (agriyakhetarpal) [inline docs/source/api/models/submodels/transport_efficiency/index.rst:9]: Could you rename the heading from `transport_efficiency` to `Transport Efficiency`? This has caused errors previously with duplicate headings in a file and its toctree

File: .virtual_documents/docs/source/examples/notebooks/models/tortuosity_models.ipynb
Comment (agriyakhetarpal) [inline .virtual_documents/docs/source/examples/notebooks/models/tortuosity_models.ipynb:1]: I think this notebook should be removed, it does not show up in the documentation and is showing up as invalid JSON

**Review by agriyakhetarpal — COMMENTED** [review]
It is a bit difficult to ascertain as to what the cause of the failure is since the failures are in the PDF build of the documentation (#3292) while the HTML and doctest builds have no issues. The logs appear to be very similar and there is hardly any difference between those for a [passing build](https://readthedocs.org/projects/pybamm/builds/22566391/) and those for [this PR (a failing build)](https://readthedocs.org/projects/pybamm/builds/22563790/), except for the keyword `failed` instead of `done` and the names of the new transport efficiency submodel entries, of course. We currently remove all notebooks-related pages in the PDF builds so I don't think the problem is in either of the two notebooks changed here. However, just to be sure and if it is possible: please add an alt-text for the images in the markdown syntax in the notebooks – not doing that was also a cause of an error I had encountered in #3043 with some of them.

I checked the references in the docstrings by matching them from the `CITATIONS.bib` file and they look fine. I can take a look at debugging the build locally but I will be a bit busy until next week, so I have left a few comments below which could help in the meantime.

File: .virtual_documents/docs/source/examples/notebooks/models/tortuosity_models.ipynb
Comment (TomTranter) [reply .virtual_documents/docs/source/examples/notebooks/models/tortuosity_models.ipynb:1]: Ok, I'll remove the file and try adding a new one. I'd like it to be included in the docs

Comment (agriyakhetarpal) [reply .virtual_documents/docs/source/examples/notebooks/models/tortuosity_models.ipynb:1]: I think there are two files of the same name though, this one is in `.virtual_documents/` and another one which is valid JSON and shows up correctly

**Comment by TomTranter** [comment]
@agriyakhetarpal I can't seem to find the offending readthedocs failure. Any suggestions?

**Comment by agriyakhetarpal** [comment]
Sorry, I forgot to give an update here but I did try this locally in the last week of November after I said I will. The error message was painfully difficult to debug, and I tried an array of approaches to fix it but resorted to leaving it be later on: moving the paths for various files, excluding all new files from this PR from the docs generation, experiments with switching to an alternative PDF builder (`rinohtype`), and more – nothing has worked so far.

However, I have been able to isolate that the problem is not with the notebooks (they are not processed during the PDF builds), and not with the directives in the `.rst` files – which suggests that this is coming from either how the file structure is distributed for the submodel or from something we missed in the docstrings for the submodel classes.

@tinosulzer had suggested enabling PDF builds when I was a GSoC participant, and so were they done soon enough, but we can look into removing them if we do not wish to include them again. Please let me know what works, in the meantime I can take another look at this on Monday.

**Review by DrSOKane — CHANGES_REQUESTED** [review]
As someone who tried implementing the user-supplied tortuosity factor myself, this is very good to see!

The problem I quickly ran into with non-Bruggeman tortuosity models is how to tell PyBaMM to disregard the tortuosity for solid-phase conductivity. With Bruggeman models it is done by specifying an exponent of 0, but this doesn't work for any of the other models. The only solutions I can think of are:
* Specify separate options for electrolyte and electrode transport efficiency, which is a lot of work for little gain
* Specify a "tortuosity for electrode conductivity" option which can only be true or false
* Parameter sets corresponding to paper that use non-Bruggeman tortuosity models should have the initial electrode conductivity adjusted accordingly
 
Furthermore, using phi = 1 - eps is incorrect for models that feature degradation. The tortuosity will decrease as the SEI grows, and LAM is not accounted for.

These are minor issues that will be easy to solve _if_ everyone can agree on which approach to use.

**Comment by TomTranter** [comment]
Hi @DrSOKane thanks for your comments and good suggestions. I think the best approach is "Specify separate options for electrolyte and electrode transport efficiency". It is perfectly valid to have different functional relationships for each phase. On the point of LAM I would need to dig into this but we should certainly account for it. We should Ideally have a volume fraction model that can account for active material, binder, SEI, non-active material (broken-off chunks) and porous space

**Comment by TomTranter** [comment]
@agriyakhetarpal thanks for he update - it certainly is a hard problem to find and fix. Maybe @martinjrobins has experience that can help or I could contact the readthedocs maintainers

**Comment by agriyakhetarpal** [comment]
I agree. Though, it doesn't look to me as something related to RTD (fails locally as well), so we would rather be better off contacting the Sphinx developers – though I can't seem to find the related issue at this moment, it had been dormant for years the last time I looked ☹️

**Comment by TomTranter** [comment]
@agriyakhetarpal I found the source of the docs errors finally. It was missing footbibliography commands on the rst files

**Comment by agriyakhetarpal** [comment]
Thanks, @TomTranter! That makes so much more sense now, I imagine it was surely not easy to find out the source of it through the unhelpful error message :)

**Comment by TomTranter** [comment]
@brosaplanella can we merge this in now? @DrSOKane made a suggestion for a code change but I don't have time to implement that right now and would rather get this is now it's passing

**Comment by brosaplanella** [comment]
I can't merge as it needs Simon's approval, given that he requested changes. @DrSOKane can you check if you are happy with it?

File: CHANGELOG.md
Comment (agriyakhetarpal) [inline CHANGELOG.md:5]:
```suggestion
```

This entry can be removed, since the `scikits.odes` solver was removed in #3932.

**Review by DrSOKane — APPROVED** [review]
I'm still not happy with the assumption that `phi = 1 - eps`. However, this is badly needed for the next release, so I will approve it. The binder can be added explicitly in a future PR.

**Pull request merged**
